assignment2/hmm_evaluation.py

```diff
@@ -0,0 +1,89 @@
+        self.insertion_errors = 0
+
+
+    def compute_errors(self):
+        gt_idx, pred_idx = 0, 0
+        #print(self.ground_truth, self.predicted)
+        while gt_idx < len(self.ground_truth) and pred_idx < len(self.predicted):
+            if self.ground_truth[gt_idx] == self.predicted[pred_idx]:
+                # No error, move both indices forward
+                # print(ground_truth[gt_idx], predicted[pred_idx])
+                gt_idx += 1
+                pred_idx += 1
+                
+            else:
+                # Substitution error
+                self.substitution_errors += 1
+                gt_idx += 1
+                pred_idx += 1
+
+        # Handle remaining ground truth words (deletions)
+        while gt_idx < len(self.ground_truth):
+            self.deletion_errors += 1
+            gt_idx += 1
+
+        # Handle remaining predicted words (insertions)
+        while pred_idx < len(self.predicted):
+            self.insertion_errors += 1
+            pred_idx += 1
+
+    def calculate_metrics(self):
+    
+        # Calculate performance metrics
+        correct_recognition_percentage = 100 * (self.N - self.substitution_errors - self.deletion_errors) / self.N
+        recognition_accuracy = 100 * (self.N - self.substitution_errors - self.deletion_errors - self.insertion_errors) / self.N
+        error_rate = 100 * (self.substitution_errors + self.deletion_errors + self.insertion_errors) / self.N
+        
+        return {
+            'Substitution Errors': self.substitution_errors,
+            'Deletion Errors': self.deletion_errors,
+            'Insertion Errors': self.insertion_errors,
+            'Percentage Correct Recognition': correct_recognition_percentage,
+            'Recognition Accuracy': recognition_accuracy,
+            'Error Rate': error_rate,
+            'Confusion Matrix': self._generate_confusion_matrix()
+        }
+
+    def _generate_confusion_matrix(self):
+        # Generates and returns the confusion matrix for the ground truth vs predicted sequences.
+        return confusion_matrix(self.ground_truth, self.predicted, labels=list(set(self.ground_truth + self.predicted)))
+
+    def print_metrics(self):
+        metrics = self.calculate_metrics()
+        
+        print(f"Substitution Errors: {metrics['Substitution Errors']}")
+        print(f"Deletion Errors: {metrics['Deletion Errors']}")
+        print(f"Insertion Errors: {metrics['Insertion Errors']}")
+        print(f"Percentage of Correct Recognition: {metrics['Percentage Correct Recognition']:.2f}%")
+        print(f"Recognition Accuracy: {metrics['Recognition Accuracy']:.2f}%")
+        print(f"Error Rate: {metrics['Error Rate']:.2f}%")
+        
+        print("\nConfusion Matrix:")
+        print(metrics['Confusion Matrix'])
+
+
+
+
+
+if __name__ == "__main__":
+    ground_truth = "heed hid head had hard hud hod hoard hood who'd heard"
+    predicted = "heed hid head hard hud hoard hood heard pad wood head"
```

Comment:
our predicted and ground truth will be compared one file to one file. 
It's better to generate a confusion matrix of N X N to see which ones get mislabeled the most (with some sorts of heatmap)

Reply:
ok i will change the metixs